rfcs/0137-nix-language-version.md

```diff
@@ -0,0 +1,650 @@
+         - `^`
+             – (-) Overlaps with derivation output syntax
+         - All of the following characters will interfere with some tooling:
+             - `!` - shells
+             - `"` - shells
+             - `#` - URLs
+             - `$` - shells
+             - `%` - URLs
+             - `&` - shells, URLs
+             - `+` - URLs
+             - `,` - natural language
+             - `/` - paths, URLs
+             - `:` - URLs
+             - `;` - shells
+             - `=` - URLs, Nix language
+             - `?` - URLs
+             - `@` - URLs, Nix language
+             - `\` - Windows paths
+     - (-) `default.nix` resolving needs specification:
+         - If for `import ./foo`, all of `./foo/default.nix{6,7,8}` exist, pick the one matching the version used by the evaluator, otherwise fail
+             - Then you'd have to specify a file using a different version explicitly
+
+   - Language versioning per "project" in a sidecar file
+       - (+) This would easily allow inheriting the language version across imports (obviating many specifications in this proposal)
+       - (-) There is currently no notion of "project" in the Nix language
+         - (-) Attempting to establish one would be a large undertaking and not immediately help solving the problem at hand
+       - (-) Cannot be introduced gradually, particularly relevant for a large codebase like Nixpkgs
+       - (-) It would require a separate language to encode project metadata such as the language version
+           - The `edition` field was [removed from the flakes schema](https://github.com/NixOS/nix/commit/e5ea01c1a8bbd328dcc576928bf3e4271cb55399) for that reason, as it not not allow distinguishing data from metadata
+           - (+) Other languages do the same (Python, Haskell, Rust, JavaScript, ...)
+             - (-) Recursive (albeit smaller) problem of managing the additional language for project metadata
+
+   </details>
+
+1. The following syntax is used for declaring the language version of a Nix expression:
+
+   ```
+   version \d*;
+   ```
+
+   This implies that if no language version is specified in a Nix file, it is written in version 6 (the version implemented in the stable release of Nix at the time of writing this RFC).
+
+   The syntax is open for bikeshedding.
+   Alternatives should be very short and self-describing.
+
+   <details><summary>Arguments</summary>
+
+   - (+) Short and fairly self-descriptive
+   - (+) Not an invasive change
+
+   </details>
+
+   <details><summary>Alternatives</summary>
+
+   - `use v\d*;`
+     - (+) Shorter
+     - (-) Does not explain much
+   - `Nix language version \d*`
+     - (+) Says it all
+     - (-) Very long
+
+   </details>
+
+1. The language version declaration is optional for bare Nix expressions, and can be specified with a parameter to the evaluator.
+   If no version is specified in a bare Nix expression, assume the most recent language version supported by the evaluator.
```

Comment:
This directly contradicts the earlier sentence saying that

> This implies that if no language version is specified in a Nix file, it is written in version 6 (the version implemented in the stable release of Nix at the time of writing this RFC).

Reply:
Yes, this is to ease use of the REPL and ad hoc commands. Should add that. Or is there any objection to that?

Reply:
Can you give an example? I guess it's unclear what "bare Nix expressions" are